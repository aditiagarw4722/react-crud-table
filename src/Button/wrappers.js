import bcc from 'bem-react-component-creator';

const { block } = bcc('crud-button');

export const Btn = block('button');

export default { Btn };
