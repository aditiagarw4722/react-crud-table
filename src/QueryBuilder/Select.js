// @flow

import React, { Component } from 'react';

type Option = {
  key: string | number,
  value: string | number,
  text: string,
};
type Props = {
  value: ?string | ?number,
  options: Array<Option>,
  placeholder: string,
  onChange: Function,
};

type State = {
  value: ?string | ?number,
};

class Select extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { value: props.value };
  }

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillReceiveProps(nextProps: Props) {
    const { value } = this.props;
    if (nextProps.value !== value) {
      this.setState({ value: nextProps.value });
    }
  }

  handleChange(evt: SyntheticEvent<HTMLSelectElement>) {
    const { onChange } = this.props;
    onChange(
      evt,
      { value: evt.currentTarget.value },
    );
  }

  render(): React$Element<'select'> {
    const { value } = this.state;
    const { placeholder, options } = this.props;
    return (
      <select onChange={this.handleChange.bind(this)}>
        <option
          value=""
          selected={!value}
        >
          {placeholder}
        </option>
        {options.map((option) => (
          <option
            key={option.key}
            value={option.value}
          >
            {option.text}
          </option>
        ))}
      </select>
    );
  }
}

export default Select;
