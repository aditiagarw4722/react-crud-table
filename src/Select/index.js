// @flow

import React, { Component } from 'react';

type Option = {
  key: string | number,
  value: string | number,
  text: string,
};
type Props = {
  value: ?string | ?number,
  options: Array<Option>,
  placeholder: string,
  onChange: Function,
};

const Select = (props: Props): React$Element<'select'> => {
  const { placeholder, options, value, onChange } = props;
  return (
    <select className="crud-table__select" onChange={onChange} value={value}>
      <option value="">{placeholder}</option>
      {options.map((option) => (
        <option key={option.key} value={option.value}>
          {option.text}
        </option>
      ))}
    </select>
  );
};

export default Select;
