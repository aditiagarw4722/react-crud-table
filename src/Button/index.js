import React from 'react';
import { Btn } from './wrappers';

const Button = (props) => <Btn {...props} />;

export default Button;
