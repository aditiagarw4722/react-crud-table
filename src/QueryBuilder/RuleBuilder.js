// @flow

import React from 'react';

import Button from '../Button';
import Select from '../Select';
import { queryValue } from '../CRUDTable/helpers';
import {
  conditionsForType,
  getDefaultConditionForType,
  inputForType,
  mapFieldsToOptions,
  isBoolean,
  isRuleComplete,
} from './helpers';
import { CONDITIONS, DEFAULT_STATE } from './constants';
import { RuleBuilder as Container } from './wrappers';

type Field = {
  value: any,
};
type Props = {
  fields: Array<Field>,
  conditionsSelectPlaceholder: string,
  fieldsSelectPlaceholder: string,
  onSave: Function,
};

type State = {
  field: any,
  type: string,
  value: any,
  label: string,
  collection: string,
  condition: string,
};

type Data = {
  value: any,
};

class RuleBuilder extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = DEFAULT_STATE;
    // $FlowFixMe
    this.handleFieldSelectChange = this.handleFieldSelectChange.bind(this);
    // $FlowFixMe
    this.save = this.save.bind(this);
  }

  handleFieldSelectChange(evt: SyntheticEvent<HTMLSelectElement>) {
    const { value } = evt.currentTarget;
    const update: State = {
      ...this.state,
      field: value,
      type: this.getType(value),
      label: this.getLabel(value),
      collection: this.getCollection(value),
    };

    update.condition = this.getDefaultCondition(value);

    if (isBoolean(update.type)) {
      update.condition = CONDITIONS.IS;
      update.value = false;
    }

    this.setState(update);
  }

  getLabel(field: any): string {
    return queryValue(this.find(field), 'label', '');
  }

  getDefaultCondition(field: any): string {
    const defaultconditionForType = getDefaultConditionForType(
      this.getType(field)
    );

    return queryValue(
      this.find(field),
      'defaultCondition',
      defaultconditionForType
    );
  }

  getCollection(field: any): string {
    return queryValue(this.find(field), 'collection', '');
  }

  getType(field: any): string {
    return queryValue(this.find(field), 'type', '');
  }

  find(field: any): ?Field {
    const { fields } = this.props;
    return fields.find((f) => f.value === field);
  }

  save() {
    const { onSave } = this.props;
    if (isRuleComplete(this.state)) {
      onSave(this.state);
      this.setState(DEFAULT_STATE);
    }
  }

  render(): React$Element<any> {
    const {
      fields,
      conditionsSelectPlaceholder,
      fieldsSelectPlaceholder,
    } = this.props;
    const { field, value, condition } = this.state;
    const type = this.getType(field);
    const input = inputForType(type, {
      value,
      onChange: (evt) => {
        this.setState({ value: evt.currentTarget.value });
      },
    });

    return (
      <Container>
        <Select
          placeholder={fieldsSelectPlaceholder}
          options={mapFieldsToOptions(fields)}
          value={field}
          onChange={this.handleFieldSelectChange}
        />
        &nbsp;&nbsp;
        {!isBoolean(type) && (
          <span>
            <Select
              placeholder={conditionsSelectPlaceholder}
              options={conditionsForType(type)}
              value={condition}
              onChange={(evt, data) => {
                this.setState({ condition: evt.currentTarget.value });
              }}
            />
            &nbsp;&nbsp;
          </span>
        )}
        {input}
        &nbsp;&nbsp;
        <Button modifiers="positive,add" onClick={this.save}>
          +
        </Button>
      </Container>
    );
  }
}

export default RuleBuilder;
