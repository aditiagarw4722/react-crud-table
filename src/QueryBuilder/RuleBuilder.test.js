import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { create } from 'react-test-renderer';

import DateTime from 'react-datetime';
import Button from '../Button';
import Select from '../Select';
import RuleBuilder from './RuleBuilder';

const renderer = new ShallowRenderer();

describe('RuleBuilder', () => {
  let props;

  beforeEach(() => {
    props = {
      fields: [
        {
          name: 'name',
          type: 'text',
          text: 'Name',
          value: 'name',
        },
        {
          name: 'active',
          type: 'boolean',
          text: 'Active',
          value: 'active',
        },
        {
          name: 'age',
          text: 'Age',
          type: 'number',
          value: 'age',
        },
        {
          name: 'date',
          text: 'Date',
          type: 'date',
          value: 'date',
        },
      ],
      conditionsSelectPlaceholder: 'Select cond',
      fieldsSelectPlaceholder: 'Select f',
      onSave: jest.fn(),
    };
  });

  it('renders as expected with the given props', () => {
    const result = renderer.render(<RuleBuilder {...props} />);
    expect(result).toMatchSnapshot();
  });

  it('can select a text field condition', () => {
    const result = create(<RuleBuilder {...props} />);

    const selects = result.root.findAllByType(Select);
    selects[0].props.onChange({
      currentTarget: {
        value: 'name',
      },
    });

    selects[1].props.onChange({
      currentTarget: {
        value: 'CONTAINS',
      },
    });

    result.root.findByType(Button).props.onClick();

    result.root.findByType('input').props.onChange({
      currentTarget: {
        value: 'Foo',
      },
    });

    result.root.findByType(Button).props.onClick();

    expect(props.onSave).toHaveBeenCalledTimes(1);
    expect(props.onSave).toHaveBeenCalledWith({
      collection: '',
      condition: 'CONTAINS',
      field: 'name',
      label: '',
      type: 'text',
      value: 'Foo',
    });
  });

  it('can select a number field condition', () => {
    const result = create(<RuleBuilder {...props} />);

    const selects = result.root.findAllByType(Select);
    selects[0].props.onChange({
      currentTarget: {
        value: 'age',
      },
    });

    selects[1].props.onChange({
      currentTarget: {
        value: 'EQUALS_TO',
      },
    });

    result.root.findByType('input').props.onChange({
      currentTarget: {
        value: 20,
      },
    });

    result.root.findByType(Button).props.onClick();

    expect(props.onSave).toHaveBeenCalledTimes(1);
    expect(props.onSave).toHaveBeenCalledWith({
      collection: '',
      condition: 'EQUALS_TO',
      field: 'age',
      label: '',
      type: 'number',
      value: 20,
    });
  });

  it('can select a boolean field condition', () => {
    const result = create(<RuleBuilder {...props} />);

    const selects = result.root.findAllByType(Select);
    selects[0].props.onChange({
      currentTarget: {
        value: 'active',
      },
    });

    result.root.findByType('input').props.onClick({
      currentTarget: {
        checked: true,
      },
    });

    result.root.findByType(Button).props.onClick();

    expect(props.onSave).toHaveBeenCalledTimes(1);
    expect(props.onSave).toHaveBeenCalledWith({
      collection: '',
      condition: 'IS',
      field: 'active',
      label: '',
      type: 'boolean',
      value: true,
    });
  });

  it('can select a date field condition', () => {
    const result = create(<RuleBuilder {...props} />);

    const selects = result.root.findAllByType(Select);
    selects[0].props.onChange({
      currentTarget: {
        value: 'date',
      },
    });

    selects[1].props.onChange({
      currentTarget: {
        value: 'GREATER_OR_EQUALS_THAN',
      },
    });

    result.root.findByType(DateTime).props.onChange({
      format: () => '2018-10-01 10:30',
    });

    result.root.findByType(Button).props.onClick();

    expect(props.onSave).toHaveBeenCalledTimes(1);
    expect(props.onSave).toHaveBeenCalledWith({
      collection: '',
      condition: 'GREATER_OR_EQUALS_THAN',
      field: 'date',
      label: '',
      type: 'date',
      value: '2018-10-01 10:30',
    });
  });
});
