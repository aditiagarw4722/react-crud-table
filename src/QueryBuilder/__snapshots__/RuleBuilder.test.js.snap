// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleBuilder renders as expected with the given props 1`] = `
<RuleBuilder>
  <Select
    onChange={[Function]}
    options={
      Array [
        Object {
          "key": "name",
          "text": undefined,
          "value": "name",
        },
        Object {
          "key": "active",
          "text": undefined,
          "value": "active",
        },
        Object {
          "key": "age",
          "text": undefined,
          "value": "age",
        },
        Object {
          "key": "date",
          "text": undefined,
          "value": "date",
        },
      ]
    }
    placeholder="Select f"
    value=""
  />
    
  <span>
    <Select
      onChange={[Function]}
      options={
        Array [
          Object {
            "key": "CONTAINS",
            "text": "Contains",
            "value": "CONTAINS",
          },
          Object {
            "key": "EQUALS_TO",
            "text": "Equals to",
            "value": "EQUALS_TO",
          },
          Object {
            "key": "BEGINS_WITH",
            "text": "Begins with",
            "value": "BEGINS_WITH",
          },
          Object {
            "key": "ENDS_WITH",
            "text": "Ends with",
            "value": "ENDS_WITH",
          },
          Object {
            "key": "IS_NOT_EQUALS_TO",
            "text": "Is not equals to",
            "value": "IS_NOT_EQUALS_TO",
          },
          Object {
            "key": "DOES_NOT_BEGIN_WITH",
            "text": " Does not begin with",
            "value": "DOES_NOT_BEGIN_WITH",
          },
          Object {
            "key": "DOES_NOT_END_WITH",
            "text": " Does not end with",
            "value": "DOES_NOT_END_WITH",
          },
          Object {
            "key": "DOES_NOT_CONTAIN",
            "text": " Does not contain",
            "value": "DOES_NOT_CONTAIN",
          },
        ]
      }
      placeholder="Select cond"
      value=""
    />
      
  </span>
  <Input
    onChange={[Function]}
    type="text"
    value=""
  />
    
  <Button
    modifiers="positive,add"
    onClick={[Function]}
  >
    +
  </Button>
</RuleBuilder>
`;
